index.d.ts

```diff
@@ -1,18 +1,24 @@
-import {Middleware, Dispatch} from "redux";
+import { Middleware, Action, AnyAction } from "redux";
 
 
-export type ThunkAction<R, S, E> = (dispatch: Dispatch<S>, getState: () => S,
-                                    extraArgument: E) => R;
+export type ThunkAction<R, S, E, A extends Action> = (
+  dispatch: ThunkDispatch<S, E, A>,
+  getState: () => S,
+  extraArgument: E
+) => R;
 
-declare module "redux" {
-  export interface Dispatch<S> {
-    <R, E>(asyncAction: ThunkAction<R, S, E>): R;
-  }
+interface ThunkDispatch<S, E, A extends Action> {
+  <T extends A>(action: T): T;
 }
 
+interface ThunkDispatch<S, E, A extends Action> {
+  <R>(asyncAction: ThunkAction<R, S, E, A>): R;
+}
+
+type ThunkMiddleware<E, S = {}, A extends Action = AnyAction> = Middleware<ThunkDispatch<S, E, A>, S, ThunkDispatch<S, E, A>>;
```

Comment:
`E` could be inferred in `withExtraArgument`. `S` could be inferred in `Middleware`. Now the problem is `A` is not inferable. I'm thinking to open a PR in https://github.com/reactjs/redux to pass action types to `Middleware`